addon/scripts/crypto.js

```diff
@@ -281,6 +278,100 @@ function decodeStorablePoint(s) {
+ * (return the xhr for testing)
+ * @param {string} version commitment version string
+ */
+function validateAndStoreTokens(url, tabId, tokens, signatures, batchProof, version) {
+    let cXhr = createVerificationXHR(url, tabId, tokens, signatures, batchProof, version);
+    // The response to the XHR will trigger cXhr.onreadystatechange()
+    cXhr.send();
+    return cXhr;
+}
+
+/**
+ * Retrieves the commitments from the GH beacon
+ * (return the xhr object for testing purposes)
+ * @param {string} version commitment version string
+ */
+function createVerificationXHR(url, tabId, tokens, signatures, batchProof, version) {
+    let xhr = new XMLHttpRequest();
+    xhr.open("GET", COMMITMENT_URL, true);
+    xhr.setRequestHeader("Content-Type", "application/json");
+    xhr.onreadystatechange = function() {
+        if (xhr.status < 300 && xhr.readyState == 4) {
+            const commitments = retrieveCommitments(xhr, version);
+            const sigPoints = getCurvePoints(signatures);
+
+            // Verify the DLEQ batch proof before handing back the usable points
+            if (!verifyProof(batchProof, tokens, sigPoints, commitments)) {
+                throw new Error("[privacy-pass]: Unable to verify DLEQ proof.")
+            }
+
+            // Store the tokens for future usage
+            storeNewTokens(tokens, sigPoints);
+
+            // Reload the page for the originally intended url
+            if (RELOAD_ON_SIGN && !url.href.includes(CHL_CAPTCHA_DOMAIN)) {
+                let captchaPath = url.pathname;
+                let pathIndex = url.href.indexOf(captchaPath);
+                let reloadUrl = url.href.substring(0, pathIndex+1);
+                setSpendFlag(url.host, true);
+                updateBrowserTab(tabId, reloadUrl);
+            }
+        }
+    };
+    return xhr;
+}
+
+/**
+ * Decodes the received curve points
+ * @param signatures encoded signed points
+ */
+function getCurvePoints(signatures) {
+    let usablePoints = [];
+    signatures.forEach(function(signature) {
+        let usablePoint = sec1DecodePoint(signature);
+        if (usablePoint == null) {
+            throw new Error("[privacy-pass]: unable to decode point " + signature + " in " + JSON.stringify(signatures));
+        }
+        usablePoints.push(usablePoint);
+    });
+    return usablePoints;
+}
+
+/**
+ * Retrieves the public commitments that are used for validating the DLEQ proof
+ * @param {XMLHttpRequest} xhr XHR for retrieving the active EC commitments
+ * @param {string} version commitment version string
+ */
+function retrieveCommitments(xhr, version) {
+    let commG;
+    let commH;
+    const respBody = xhr.responseText;
+    let resp = JSON.parse(respBody);
+    let comms = resp[COMMITMENTS_KEY];
+    if (comms) {
+        if (DEV) {
+            commG = comms["dev"]["G"];
+            commH = comms["dev"]["H"];
+        } else if (version) {
+            commG = comms[version]["G"];
```

Comment:
Q: Does version can be `"dev"` string? Is there a checking/validation process for the version parameter?

Reply:
Yes it could be dev. Good point about the validation. I've put a null check outside of the function to check that the commitments are actually defined after retrieval.